src/qutip_qip/decompose/_utility.py

```diff
@@ -29,3 +32,469 @@ def check_gate(gate, num_qubits):
+                if k == '0':
+                    back_new_target = back_target_info[j]
+                    back_new_target_info.append(back_new_target)
+                    # change control value to be 1 if it is 0
+                    backward_gray_code[key]['control_value ='] = [
+                        '1']*back_num_control_qubits
+
+            if len(back_new_target_info) != 0:
+                # returns info about which two-level array gate decomposition
+                # needs pauli x to correct ctrl_value = 0
+                back_array_gate_info_with_control_0.append(
+                    [key, back_new_target_info])
+
+        # create pauli arrays
+        back_full_left_pauli_gate_dict = {}
+        back_full_right_pauli_gate_dict = {}
+        for i in range(len(back_array_gate_info_with_control_0)):
+            left_pauli_array = []
+            right_pauli_array = []
+            gate_targets = back_array_gate_info_with_control_0[i][1]
+            for j in range(len(gate_targets)):
+                target_qubit = gate_targets[j]
+                paulix = Gate("X", targets=target_qubit)
+                left_pauli_array.append(paulix)
+            key = back_gate_keys[i]
+            back_full_left_pauli_gate_dict[key] = left_pauli_array
+            # the pauli gate order is reversed when applied from the right
+            right_pauli_array = left_pauli_array[::-1]
+            back_full_right_pauli_gate_dict[key] = right_pauli_array
+
+        # create full forward gate dict with pauli gates
+        back_full_gate_dict = {}
+        for i in range(len(backward_gray_code)):
+            if len(back_array_gate_info_with_control_0) == len(
+                    backward_gray_code):
+                key = back_gate_keys[i]
+                left_pauli = back_full_left_pauli_gate_dict[key]
+                right_pauli = back_full_right_pauli_gate_dict[key]
+                gray_code_with_ctrl_1 = backward_gray_code[key]
+                back_full_gate_i_info = [
+                    left_pauli, gray_code_with_ctrl_1, right_pauli]
+                back_full_gate_dict[key] = back_full_gate_i_info
+            elif len(for_array_gate_info_with_control_0) < len(
+                    forward_gray_code):
+                key = back_gate_keys[i]
+                if key in gate_keys:
+                    left_pauli = back_full_left_pauli_gate_dict[key]
+                    right_pauli = back_full_right_pauli_gate_dict[key]
+                    gray_code_with_ctrl_1 = backward_gray_code[key]
+                    back_full_gate_i_info = [
+                        left_pauli, gray_code_with_ctrl_1, right_pauli]
+                    back_full_gate_dict[key] = back_full_gate_i_info
+                else:
+                    back_full_gate_dict[key] = backward_gray_code[key]
+
+        # calculated lengths
+        len_forward_gate_info = len(full_gate_dict)
+        len_backward_gate_info = len(back_full_gate_dict)
```

Comment:
Same as above comment - if there's more than one step then the list of gates can be reversed for the backward gray code gates. There's no need to iterate over forward gray code info to find the last gate object - a new function will output it separately.